File: components/toast/ios/CiataToast.swift
import SwiftUI

/// Implementação experimental do CMP-0012 Toast para SwiftUI.
public struct CiataToast: View {
    private let message: String
    private let actionLabel: String?
    private let action: (() -> Void)?

    public init(
        message: String,
        actionLabel: String? = nil,
        action: (() -> Void)? = nil
    ) {
        let normalizedMessage = message.trimmingCharacters(in: .whitespacesAndNewlines)
        let normalizedActionLabel = actionLabel?.trimmingCharacters(in: .whitespacesAndNewlines)

        precondition(!normalizedMessage.isEmpty, "message não pode ser vazio")
        precondition((normalizedActionLabel == nil) == (action == nil), "actionLabel e action devem ser fornecidos juntos")
        precondition(normalizedActionLabel == nil || !normalizedActionLabel!.isEmpty, "actionLabel não pode ser vazio")

        self.message = normalizedMessage
        self.actionLabel = normalizedActionLabel
        self.action = action
    }

    public var body: some View {
        HStack(alignment: .top, spacing: 12) {
            Text(message)
                .frame(maxWidth: .infinity, alignment: .leading)
            if let actionLabel, let action {
                Button(actionLabel, action: action)
            }
        }
        .padding()
        .accessibilityElement(children: .contain)
    }
}

File: components/toast/flutter/ciata_toast.dart
import 'package:flutter/material.dart';

/// Implementação experimental do CMP-0012 Toast para Flutter.
class CiataToast extends StatelessWidget {
  CiataToast({
    super.key,
    required String message,
    String? actionLabel,
    this.onAction,
  })  : message = message.trim(),
        actionLabel = actionLabel?.trim() {
    if (this.message.isEmpty) {
      throw ArgumentError.value(message, 'message', 'não pode ser vazio');
    }
    if ((this.actionLabel == null) != (onAction == null)) {
      throw ArgumentError(
        'actionLabel e onAction devem ser fornecidos juntos',
      );
    }
    if (this.actionLabel != null && this.actionLabel!.isEmpty) {
      throw ArgumentError.value(
        actionLabel,
        'actionLabel',
        'não pode ser vazio',
      );
    }
  }

  final String message;
  final String? actionLabel;
  final VoidCallback? onAction;

  @override
  Widget build(BuildContext context) {
    return Semantics(
      liveRegion: true,
      container: true,
      child: Material(
        child: Padding(
          padding: const EdgeInsets.all(16),
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Expanded(child: Text(message)),
              if (actionLabel != null && onAction != null) ...[
                const SizedBox(width: 12),
                TextButton(onPressed: onAction, child: Text(actionLabel!)),
              ],
            ],
          ),
        ),
      ),
    );
  }
}
